### PR TITLE
Extract credential handler and apply to all the http client

### PR DESCRIPTION
**Pull request by 928PJY**
[AB#336926](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/336926)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6929)

### CODE REVIEW

**Review by yufeih — CHANGES_REQUESTED** [review]
We may not need this change if using [Polly Retry to refresh authorization](https://github.com/App-vNext/Polly/wiki/Retry#retry-to-refresh-authorization)

**Comment by herohua** [comment]
>     internal sealed class GitHubAccessor

should we implement `IDisposable`?

---
Refers to: src/docfx/lib/github/GitHubAccessor.cs:22 in c6ae779. [](commit_id = c6ae779dce083f233c2b34b51bb60612aca72438, deletion_comment = False)

File: src/docfx/config/ops/CredentialHandler.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Identity;
+using Azure.Security.KeyVault.Secrets;
+
+namespace Microsoft.Docs.Build
+{
+    internal class CredentialHandler : DelegatingHandler
```
Comment (herohua) [inline src/docfx/config/ops/CredentialHandler.cs:15]: >CredentialHandler [](start = 19, length = 17)

shall we call this `OpsCredentialHandler`?

File: src/docfx/config/ops/CredentialHandler.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                if (TryExtractDocsEnvironmentFromUrl(request.RequestUri?.AbsoluteUri, out _))
+                {
+                    return await HandleOpsRequest(request, cancellationToken);
+                }
+
+                return await base.SendAsync(request, cancellationToken);
+            }
+        }
+
+        public static string BuildServiceEndpoint(DocsEnvironment? environment = null)
```
Comment (herohua) [inline src/docfx/config/ops/CredentialHandler.cs:61]: >BuildServiceEndpoint [](start = 29, length = 20)

It is kind of weird to me that `CredentialManager` exposes a method called `BuildServiceEndpoint`. Shall we consider extracting environment and endpoint info into a separate class, e.g. named `OpsEnvironments`?

File: src/docfx/config/PreloadConfig.cs [hunk 3]
```diff
@@ -57,27 +57,16 @@ internal class PreloadConfig
         /// </summary>
         public string? DocsRepositoryOwnerName { get; init; }
 
-        public Action<HttpRequestMessage> GetCredentialProvider()
+        public CredentialHandler GetCredentialHandler()
         {
-            var rules = Http.OrderByDescending(pair => pair.Key, StringComparer.Ordinal).ToArray();
-
-            return message =>
+            return new Lazy<CredentialHandler>(() =>
```
Comment (herohua) [inline src/docfx/config/PreloadConfig.cs:62]: >new Lazy<CredentialHandler> [](start = 19, length = 27)

what is the meaning for this `Lazy` as we immediately invoke `.Value`?

File: test/docfx.Test/build/OpsConfigAdapterTest.cs [hunk 4]
```diff
@@ -36,7 +37,21 @@ public static async Task AdaptOpsServiceConfig(string url, string expectedJson)
                 return;
             }
 
-            var accessor = new OpsAccessor(null, request => request.Headers.Add("X-OP-BuildUserToken", token));
+            var credentialHanlder = new CredentialHandler(
```
Comment (herohua) [inline test/docfx.Test/build/OpsConfigAdapterTest.cs:40]: >Hanlder [](start = 26, length = 7)

typo

File: src/docfx/restore/CredentialHandler.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal class CredentialHandler : DelegatingHandler
+    {
+        private readonly List<KeyValuePair<string, HttpConfig>> _credentials;
+
+        public CredentialHandler(List<KeyValuePair<string, HttpConfig>> credentials, HttpMessageHandler innerHandler)
```
Comment (herohua) [inline src/docfx/restore/CredentialHandler.cs:16]: >List [](start = 33, length = 4)

`IReadOnlyList`

File: src/docfx/restore/CredentialHandler.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal class CredentialHandler : DelegatingHandler
+    {
+        private readonly List<KeyValuePair<string, HttpConfig>> _credentials;
```
Comment (herohua) [inline src/docfx/restore/CredentialHandler.cs:14]: >List [](start = 25, length = 4)

`IReadOnlyList`

File: src/docfx/restore/CredentialHandler.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+        public CredentialHandler(List<KeyValuePair<string, HttpConfig>> credentials, HttpMessageHandler innerHandler)
+            : base(innerHandler)
+        {
+            _credentials = credentials;
+        }
+
+        public CredentialHandler Create(HttpMessageHandler innerHandler)
```
Comment (herohua) [inline src/docfx/restore/CredentialHandler.cs:22]: >Create [](start = 33, length = 6)

this looks weird - why do we want to new a `CredentialHandler` with exiting credentials

File: src/docfx/config/ops/OpsCredentialHandler.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Identity;
+using Azure.Security.KeyVault.Secrets;
+
+namespace Microsoft.Docs.Build
+{
+    internal class OpsCredentialHandler : DelegatingHandler
```
Comment (herohua) [inline src/docfx/config/ops/OpsCredentialHandler.cs:14]: >OpsCredentialHandler [](start = 19, length = 20)

should `OpsCredentialHandler` be a `CredentialHandler`?

File: src/docfx/config/ops/OpsCredentialHandler.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            // Default header which allows fallback to public data when credential is not provided.
+            request.Headers.TryAddWithoutValidation("X-OP-FallbackToPublicData", true.ToString());
+            await FillOpsToken(request);
+            return await base.SendAsync(request, cancellationToken);
+        }
+
+        private static async Task FillOpsToken(HttpRequestMessage request, DocsEnvironment? environment = null)
+        {
+            if (IsSameDocsEnvironmentRequest(request))
```
Comment (yufeih) [inline src/docfx/config/ops/OpsCredentialHandler.cs:38]: This line makes me want to get away from using `DelegatingHandler` for middlewares. Its contract makes passing variables like `environment` here so much harder.

All we need is to turn the existing credential provider from `Action<HttpRequestMessage>` to:

```csharp
delegate Task<HttpResponseMessage> CredentialProvider(HttpRequestMessage request, Func<HttpRequestMessage, Task<HttpResponseMessage>> next);
```

Then you only need to change places that calls `_credentialProvider?.Invoke(request);`

**Review by yufeih — CHANGES_REQUESTED** [review]

File: src/docfx/config/ops/CredentialHandler.cs (hunk 1 above)
Comment (928PJY) [reply src/docfx/config/ops/CredentialHandler.cs]: it handles all the credentials not only ops

File: src/docfx/config/ops/CredentialHandler.cs (hunk 2 above)
Comment (928PJY) [reply src/docfx/config/ops/CredentialHandler.cs]: moved back to opsAccessor

File: src/docfx/restore/CredentialHandler.cs (hunk 7 above)
Comment (928PJY) [reply src/docfx/restore/CredentialHandler.cs:22]: it is to create `credentialHandler` with a different `innerHandler`

File: src/docfx/config/ops/OpsCredentialHandler.cs (hunk 8 above)
Comment (928PJY) [reply src/docfx/config/ops/OpsCredentialHandler.cs:14]: They handling different kind of credentials, opsCredentialHanlder doesn't inherit from the `CredentialHandler`, Rename `CredentialHandler` may avoid the confusion, but I don't have a better idea

File: src/docfx/config/ops/OpsCredentialHandler.cs (hunk 9 above)
Comment (928PJY) [reply src/docfx/config/ops/OpsCredentialHandler.cs]: Then the _credentialProvider.invoke need to be called everywhere, and it looks wired to keep a state in the delegate when we need to keep the refreshedToken.

File: src/docfx/config/PreloadConfig.cs (hunk 3 above)
Comment (928PJY) [reply src/docfx/config/PreloadConfig.cs:62]: To get the same instance when it gets called multiple times

File: src/docfx/lib/HttpRequestUtility.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Net.Http;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class HttpRequestUtility
```
Comment (live1206) [inline src/docfx/lib/HttpRequestUtility.cs:8]: not in use?

File: src/docfx/restore/CredentialProvider.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+        public Dictionary<string, string> GetCredentials(HttpRequestMessage request)
+        {
+            var credentials = new Dictionary<string, string>();
+            if (request.RequestUri?.ToString() is string url)
+            {
+                GetCredentialsCore(credentials, url, _credentials);
+                GetCredentialsCore(credentials, url, OpsAccessor.FallBackCredentials);
+            }
+            return credentials;
```
Comment (live1206) [inline src/docfx/restore/CredentialProvider.cs:27]: should this be cached? Not constructed per request.

File: src/docfx/lib/HttpRequestUtility.cs (hunk 10 above)
Comment (928PJY) [reply src/docfx/lib/HttpRequestUtility.cs:8]: it is an extension of `HttpRequestMessage `, used in `CredentialHandler.cs`.

**Review by live1206 — APPROVED** [review]

**Pull request merged**